integration/localnet/bootstrap.go

```diff
@@ -584,3 +607,159 @@ func openAndTruncate(filename string) (*os.File, error) {
+	networkKey, err := utils.GenerateUnstakedNetworkingKey(networkSeed)
+	if err != nil {
+		panic(err)
+	}
+
+	// hex encode
+	keyBytes := networkKey.Encode()
+	output := make([]byte, hex.EncodedLen(len(keyBytes)))
+	hex.Encode(output, keyBytes)
+
+	// write to file
+	outputFile := fmt.Sprintf("%s/private-root-information/%s_key", BootstrapDir, observerName)
+	err = ioutil.WriteFile(outputFile, output, 0600)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func prepareObserverService(i int, observerName string, agPublicKey string, profilerDir string, dataDir string) Service {
+	observerService := Service{
+		Image: fmt.Sprintf("localnet-%s", DefaultObserverName),
+		Command: []string{
+			fmt.Sprintf("--staked=false"),
+			fmt.Sprintf("--bootstrap-node-addresses=%s:%d", DefaultAccessGatewayName, AccessPubNetworkPort),
+			fmt.Sprintf("--bootstrap-node-public-keys=%s", agPublicKey),
+			fmt.Sprintf("--observer-networking-key-path=/bootstrap/private-root-information/%s_key", observerName),
+			fmt.Sprintf("--bind=0.0.0.0:0"),
+			fmt.Sprintf("--rpc-addr=%s:%d", observerName, RPCPort),
+			fmt.Sprintf("--secure-rpc-addr=%s:%d", observerName, SecuredRPCPort),
+			fmt.Sprintf("--http-addr=%s:%d", observerName, HTTPPort),
+			fmt.Sprintf("--collection-ingress-port=%d", RPCPort),
+			"--log-tx-time-to-finalized",
+			"--log-tx-time-to-executed",
+			"--log-tx-time-to-finalized-executed",
+			"--bootstrapdir=/bootstrap",
+			"--datadir=/data/protocol",
+			"--secretsdir=/data/secret",
+			"--loglevel=DEBUG",
+			fmt.Sprintf("--profiler-enabled=%t", true),
+			fmt.Sprintf("--tracer-enabled=%t", false),
+			"--profiler-dir=/profiler",
+			"--profiler-interval=2m",
+		},
+		Volumes: []string{
+			fmt.Sprintf("%s:/bootstrap:z", BootstrapDir),
+			fmt.Sprintf("%s:/profiler:z", profilerDir),
+			fmt.Sprintf("%s:/data:z", dataDir),
+		},
+		Environment: []string{
+			"JAEGER_AGENT_HOST=jaeger",
+			"JAEGER_AGENT_PORT=6831",
+			"BINSTAT_ENABLE",
+			"BINSTAT_LEN_WHAT",
+			"BINSTAT_DMP_NAME",
+			"BINSTAT_DMP_PATH",
+		},
+	}
+	// observer services rely on the access gateway
+	observerService.DependsOn = []string{
+		fmt.Sprintf(DefaultAccessGatewayName),
+	}
+	observerService.Ports = []string{
+		// Flow API ports come in pairs, open and secure. While the guest port is always
+		// the same from the guest's perspective, the host port numbering accounts for the presence
+		// of multiple pairs of listeners on the host to avoid port collisions. Observer listener pairs
+		// are numbered just after the Access listeners on the host network by prior convention
+		fmt.Sprintf("%d:%d", (accessCount*2)+FlowAPIPort+(2*i), RPCPort),
+		fmt.Sprintf("%d:%d", (accessCount*2)+FlowAPIPort+(2*i)+1, SecuredRPCPort),
+	}
+	return observerService
+}
+
+func prepareObserverServices(dockerServices Services, flowNodeContainerConfigs []testnet.ContainerConfig) Services {
+	agPublicKey, err := getAccessGatewayPublicKey(flowNodeContainerConfigs)
```

Comment:
nit: you can do this lookup after the validity checks

Reply:
Updated, thank you.